integ tests for exporter n reader

KishoreKicha14 opened the pull request:
### Description
This PR improves the QueryInsightsExporterIT test by adding the following updates:

1. Create Document 
2. Configure Top Queries Exporter
3. Perform Search
4. Set Latency Window Size
5. Check the local index creation 
6. Disable and Re-enable Exporter
7. Set Exporter to Debug Mode

### Issues Resolved
Closes [#233]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ansjcy commented:
Thanks! The cases make the exporter tests much more complete. I have several suggestions:
- In checkLocalIndices, Can we actually check the content of the indices? 
- Since we introduced the default index template in https://github.com/opensearch-project/query-insights/pull/254, can we also test if the index template is successfully created? 
- Can we also test get the historical data from reader? i.e. Call the top queries API with `from` and `end` to see if you can get the same data.

KishoreKicha14 commented:
> Thanks! These cases make the exporter tests much more complete. I have several suggestions:
> 
> * In checkLocalIndices, Can we actually check the content of the indices?
> * Since we introduced the default index template in [Add default index template for query insights local index #254](https://github.com/opensearch-project/query-insights/pull/254), can we also test if the index template is successfully created?
> * Can we also test get the historical data from reader? i.e. Call the top queries API with `from` and `end` to see if you can get the same data.

Added these changes and also added a step to check topqueries-* is green

deshsidd commented:
Triggering the tests again to make sure they are not flaky

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsExporterIT.java:
```diff
@@ -35,29 +36,192 @@ public void testQueryInsightsExporterSettings() throws IOException {
+            "{ \"persistent\" : { \"search.insights.top_queries.exporter.type\" : local_index, \"search.insights.top_queries.exporter.config.index\" : \"1a2b\" } }" };
     }
+
+    private void createIndexTemplate() throws IOException {
```
Removed

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsExporterIT.java:
```diff
@@ -35,29 +36,192 @@ public void testQueryInsightsExporterSettings() throws IOException {
         Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+        setLatencyWindowSize("1m");
+        createIndexTemplate();
+        waitForWindowToPass(10);
```
Added because giving some time to create the document

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsExporterIT.java:
```diff
@@ -35,29 +36,192 @@ public void testQueryInsightsExporterSettings() throws IOException {
+        checkQueryInsightsIndexTemplate();
+        disableLocalIndexExporter();
+        reEnableLocalIndexExporter();
+        setLocalIndexToDebug();
```
To whether is able to change from local to debug

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsExporterIT.java:
```diff
@@ -35,29 +36,192 @@ public void testQueryInsightsExporterSettings() throws IOException {
+        waitForWindowToPass(10);
+        performSearch();
+
+        waitForWindowToPass(70);
```
Add Explanation

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/reader/QueryInsightsReaderIT.java:
```diff
@@ -0,0 +1,155 @@
+import org.opensearch.plugin.insights.QueryInsightsRestTestCase;
+
+public class QueryInsightsReaderIT extends QueryInsightsRestTestCase {
+    public void testQueryInsightsReaderSettings() throws IOException, InterruptedException {
```
Changed to testQueryInsightsHistoricalTopQueriesRead

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/reader/QueryInsightsReaderIT.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private void cleanup(String fullIndexName) throws IOException, InterruptedException {
+        Thread.sleep(10000);
```
I think 10 seconds is needed for reader otherwise it is not able to find the Local Index (only for Reader)

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/reader/QueryInsightsReaderIT.java:
```diff
@@ -0,0 +1,155 @@
+
+        try {
+            client().performRequest(new Request("DELETE", "/top_queries"));
+        } catch (ResponseException ignored) {}
```
Added Logger

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/reader/QueryInsightsReaderIT.java:
```diff
@@ -0,0 +1,155 @@
+        assertTrue("Expected search results for title='Test Document'", responseContent.contains("\"Test Document\""));
+    }
+
+    private void createDocument() throws IOException {
```
Move all common utility methods into QueryInsightsRestTestCase

ansjcy commented:
@KishoreKicha14 You need to rebase main to pick up some important fixes from Core: https://github.com/opensearch-project/query-insights/pull/296

KishoreKicha14 commented on src/test/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsExporterIT.java:
```diff
@@ -9,21 +9,22 @@
 
-/** Rest Action tests for query  */
+/** Rest Action tests for query */
 public class QueryInsightsExporterIT extends QueryInsightsRestTestCase {
```
Updated these changes

ansjcy commented:
Thanks @KishoreKicha14 for the change! Just went through it and the tests looks pretty good to me.
